### DISCUSSION TITLE
Running ->store() on S3 uploads is slow

### DISCUSSION THREAD

**Discussion by bkilshaw** [livewire/livewire#4003] [1. Help] [answered]
So I'm using Filepond to upload files directly to S3 which is working great. The temporary files upload directly to S3 as expected. The problem I'm running into is when the user click save and I have to store the files.

```php
        foreach($pictures as $picture) {
            $filename = $picture->store('/', 'pictures_originals');

            if($filename){
                $pictureInstance = $checkin->pictures()->create([
                    'original_filename' => $filename
                ]);

                ProcessPicture::dispatch($pictureInstance);
            }
        }
```

The issue here is that because `$picture` is already stored on S3, and because the `pictures_original` filesystem is also a S3 bucket, `$picture->store('/', 'pictures_originals')` causes PHP to download the file locally and upload it again which is very slow, especially with larger files.

My goal when uploading directly to S3 was to speed up the save request in the first place. I don't want the user to have to wait for me to move files around until I send them a response.

What would be perfect is if `$picture->store()` simply moved the file from `livewire-tmp` to `pictures_originals` instead of downloading/uploading a new copy.

Anyway to make this work? Am I approaching this the wrong way?

For reference, here's my component and CheckInService: https://gist.github.com/bkilshaw/7174a24ee5617e6fa3fd6a015c76ff90

Thanks!

**Comment by bkilshaw** [comment, marked as answer]
Alright so after more investigation it looks like:

`$file->store()` will download a copy from S3 locally, and upload it back to the new destination. Unacceptable to have this as part of the request lifecycle since it's so slow. Might be workable with tiny files but you'll be punished if you're working with files that are more than a couple MB.

`$file->move()` can move a file between "directories" on the **same** S3 bucket. It works by running the copyObject() command directly on S3 and then delete() once it's complete. This takes about 350-400ms per object. If you're doing a single file you might be able to get away with it in the request life cycle but multiple files add up fast. In my case the users are uploading 6-8 pictures so this would result in a 3-4second delay. Again, unacceptable.

I think the solve is to move all of this into a queued job. The [livewire:configure-s3-upload-cleanup](https://github.com/livewire/livewire/blob/master/src/Commands/S3CleanupCommand.php#L29) configures your buckets to delete files older than 24 hours. That gives us plenty of time for a queued job to move the file out of temporary storage to a permanent destination. Process:

1. Use Filepond to upload files direcitly to S3 into a "livewire-tmp" bucket
2. When the user saves the records, we save the `$file->path()` to the record
3. We have a queued job that runs `$file->move($file->temporary_path, 'final_destination/'.$file->name)` which will move the file from the livewire-tmp directory to it's final destination
4. Any uploaded files that haven't been moved after 24 hours will get cleaned up by the S3 expiration rules

The one potential issue to keep in mind is that your file will not be available at the end of the request since it's still sitting in a temporary directory waiting to be processed by the queued job so make sure you handle that situation in your UI. I'll be handling it by showing a placeholder that the file exists but label it as "processing" until the jobs complete.

**Reply by tnylea** [reply]
Ran into the same situation and had to use that `file->move` functionality.

If this helps anyone else here, this is how I got it to work when using local storage:

```
Storage::move(FileUploadConfiguration::directory() . '/' . $this->file->getFilename(), 'new_folder/' . $this->file->getFilename());
```

> Make sure to include the `use Livewire\FileUploadConfiguration;` and `use Illuminate\Support\Facades\Storage;` in your class.

Then, to get it to work with s3, I used a similar approach as yours:

```
Storage::move($this->file->getRealPath(), 'final_destination/' . $this->file->getFilename());
```

It would be cool to have this documented, or it would be nice to override the `move` function in the `Livewire\TemporaryUploadedFile.php` This way we could simply do something like the following:

```
$this->file->move('folder_name');

// or to save it with a different name:
$this->file->moveAs('folder_name', 'filename');
```

Keep in mind that instead of using the `storePubliclyAs`, you can always set the [File Visibility](https://laravel.com/docs/filesystem#file-visibility) of the file after you move it:

```
Storage::setVisibility('filename.jpg', 'public');
```

I may try and do a PR later to get that simple move method added to the `Livewire\TemporaryUploadedFile.php` 🕺

**Reply by lbattaglioli2000** [reply]
my hero <3

**Reply by younus93** [reply]
Absolute life saver. I had a filepond component which uploads multiple files. I use s3 for both temp & final storage. Large files (>300mb) kept hitting 100% and being stuck there indefinitely. It took between 2 to 30 mins for a file to get uploaded successfully. I never realized that this was happening behind the scenes. 

I changed the code to Storage::disk('s3)->move($temp->path, $finalPath) and it reduced the time to upload time + a few seconds. 

Thank you for pointing this out.
